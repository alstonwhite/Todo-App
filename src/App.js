import React, { useState, useEffect } from "react";
import {useSelector, useDispatch} from "react-redux";
import "./styles.css";

import store from './redux/store'
import TodoList from "./components/TodoList";
import TodoDetail from "./components/TodoDetail";
import {getTasksServer} from "./ServerCall";
import {updateTaskItem, addTaskItem, deleteTaskItem, completeTaskItem} from './state';
import {addTaskR, selectTaskR, updateTaskR, deleteTaskR, completeTaskR} from './redux/actions';

export default function App() {
  // const [items, setItems] = useState(undefined);
  // const [activeItem, setActiveItem] = useState(undefined);
  const itemsR = useSelector(state => state.items);
  const activeItemR = useSelector(state => state.activeItem);
  const dispatch = useDispatch();
  // const state = store.getState();

  async function addTask() {
    // const nextItems = await addTaskItem(items);
    // setItems(nextItems);
    dispatch(addTaskR(itemsR));
  }

  function selectTask(item) {
    // setActiveItem(item);
    dispatch(selectTaskR(item));
  }

  function updateTask(item, title, body) {
    // const nextItems = updateTaskItem(items, item, title, body);
    // setActiveItem(nextItems.find(obj => obj._id === item._id));
    // setItems(nextItems);
    dispatch(updateTaskR(itemsR, item, title, body));
    dispatch(selectTaskR(itemsR.find(obj => obj._id === item._id)));
  }

  function deleteTask(item) {
    // const nextItems = deleteTaskItem(items, item);
    // setItems(nextItems);
    // setActiveItem(nextItems[0]);
    dispatch(deleteTaskR(itemsR, item));
    dispatch(selectTaskR(itemsR[0]));
  }

  function completeTask(item) {
    // const nextItems = completeTaskItem(items, item)
    // setItems(nextItems);
    dispatch(completeTaskR(itemsR, item));
  }

  useEffect(() => {
      getTasksServer().then(tasks => {
        // setItems(tasks);
        dispatch(updateTaskR(tasks));
        // setActiveItem(tasks[0]);
        dispatch(selectTaskR(tasks[0])); 
      });
    }, []);

  return (
    <div className="wrapper">
      <TodoList
        // items={items}
        items={itemsR}
        onAdd={addTask}
        onSelect={selectTask}
        onDelete={deleteTask}
      />
      <TodoDetail
        // active={activeItem}
        active={activeItemR}
        onUpdateTask={updateTask}
        onComplete={completeTask}
        onDelete={deleteTask}
      />
    </div>
  );
}
