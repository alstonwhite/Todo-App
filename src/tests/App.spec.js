import React from "react";
import renderer from "react-test-renderer";

import App from "../App";

const addTask = jest.fn();
const selectTask = jest.fn();
const deleteTask = jest.fn();
const updateTask = jest.fn();
const completeTask = jest.fn();

it("Renders correctly (no items)", () => {
  const items = [];
  const tree = renderer
  .create(
    <App
      items={items}
      onAdd={addTask}
      onSelect={selectTask}
      onDelete={deleteTask}
      active={items[0]}
      onUpdateTask={updateTask}
      onComplete={completeTask}
    />
  );
  expect(tree.toJSON()).toMatchSnapshot();
});

it("Renders correctly (nmultiple items)", () => {
  const items = [
    {
      _id: 1,
      title: "New task...",
      status: "new",
      body: "Add notes..."
    },
    {
      _id: 2,
      title: "New task...",
      status: "new",
      body: "Add notes..."
    },
    {
      _id: 3,
      title: "New task...",
      status: "new",
      body: "Add notes..."
    }
  ];
  const tree = renderer
  .create(
    <App
      items={items}
      onAdd={addTask}
      onSelect={selectTask}
      onDelete={deleteTask}
      active={items[0]}
      onUpdateTask={updateTask}
      onComplete={completeTask}
    />
  );
  expect(tree.toJSON()).toMatchSnapshot();
});


// Select task updates activeItem
// Delete buttons removes item from list (arr length decreases)
// Delete buttons removes correct item (by ._id)
// Delete buttons for active item sets first item as active
// Complete button 

// Key error with ul / li ??
