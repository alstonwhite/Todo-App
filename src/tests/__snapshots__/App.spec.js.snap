// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly (nmultiple items) 1`] = `
<div
  className="wrapper"
>
  <div
    className="app-column"
  >
    <div
      className="header app-header"
    >
      <div
        className="app-title"
      >
        To-Do App
      </div>
      <button
        className="button button-medium button-add"
        onClick={[Function]}
        value="Add"
      >
        Add
      </button>
    </div>
    <div
      className="task-list"
    >
      <ul />
    </div>
  </div>
  <div
    className="app-column"
  >
    <div
      className="header task-header"
    >
      <input
        className="title-input"
        onChange={[Function]}
        type="text"
        value=""
      />
      <button
        className="button button-large button-complete"
        onClick={[Function]}
      >
        Complete
      </button>
       
      <button
        className="button button-large button-delete"
        onClick={[Function]}
      >
        Remove
      </button>
       
    </div>
    <div
      className="task-body"
    >
      <textarea
        className="body-text"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`Renders correctly (no items) 1`] = `
<div
  className="wrapper"
>
  <div
    className="app-column"
  >
    <div
      className="header app-header"
    >
      <div
        className="app-title"
      >
        To-Do App
      </div>
      <button
        className="button button-medium button-add"
        onClick={[Function]}
        value="Add"
      >
        Add
      </button>
    </div>
    <div
      className="task-list"
    >
      <ul />
    </div>
  </div>
  <div
    className="app-column"
  >
    <div
      className="header task-header"
    >
      <input
        className="title-input"
        onChange={[Function]}
        type="text"
        value=""
      />
      <button
        className="button button-large button-complete"
        onClick={[Function]}
      >
        Complete
      </button>
       
      <button
        className="button button-large button-delete"
        onClick={[Function]}
      >
        Remove
      </button>
       
    </div>
    <div
      className="task-body"
    >
      <textarea
        className="body-text"
        onChange={[Function]}
        type="text"
        value=""
      />
    </div>
  </div>
</div>
`;
